File: generate_explanations.py
#!/usr/bin/env python3
"""
Gemini API で全問題に detailed_explanation を生成。
progress.json に進捗保存。中断しても再実行で続きから。
"""
import json, sys, time, os, urllib.request, urllib.error

PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data.js')
PROGRESS = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'progress.json')
API_KEY = 'AIzaSyCJgfa_csIOYy8MzAPTF2OtP1wH43avDgc'
MODEL = 'gemini-2.5-flash-lite'
WAIT = 4
MAX_RETRIES = 10

def load_questions():
    with open(PATH, 'r', encoding='utf-8') as f:
        c = f.read()
    return json.loads(c[c.index('['):].rstrip().rstrip(';'))

def save_data_js(qs):
    with open(PATH, 'w', encoding='utf-8') as f:
        f.write('const QUESTIONS = ' + json.dumps(qs, ensure_ascii=False, separators=(',', ': ')) + ';\n')

def load_progress():
    if os.path.exists(PROGRESS):
        with open(PROGRESS, 'r', encoding='utf-8') as f:
            return json.load(f)
    return {}

def save_progress(p):
    with open(PROGRESS, 'w', encoding='utf-8') as f:
        json.dump(p, f, ensure_ascii=False)

def build_prompt(q):
    ct = '\n'.join(f"{c['letter']}. {c['text']}" for c in q.get('choices', []))
    return f"""あなたはCompTIA Security+ (SY0-701) の試験対策の専門家です。
以下の問題について、IT初心者でも理解できるように詳しく日本語で解説してください。

【問題】
{q['question']}

【選択肢】
{ct}

【正解】
{q['answer']}

以下の4セクション構成で解説してください。各セクションに見出し（##）をつけてください。

## 正解の解説
- なぜこの答えが正解なのか、技術的な根拠と背景を詳しく説明
- 具体的な使用例やシナリオがあれば含める

## 不正解の選択肢の分析
- 各不正解の選択肢について、それぞれなぜ不正解なのかを個別に説明
- 正解の選択肢との違いを明確に
- ひっかけポイントがあれば指摘

## 関連する重要概念
- この問題に関連するセキュリティの概念・用語・技術を説明
- 関連する他の技術やプロトコルとの関係性
- 実務での具体的な適用例

## 試験対策のポイント
- この問題で問われている知識の要点をまとめる
- 類似問題が出た場合の判断基準
- 覚えるべきキーワードや区別のコツ

丁寧で分かりやすい解説をお願いします。"""

def call_gemini(prompt):
    url = f'https://generativelanguage.googleapis.com/v1beta/models/{MODEL}:generateContent?key={API_KEY}'
    payload = json.dumps({
        'contents': [{'parts': [{'text': prompt}]}],
        'generationConfig': {'temperature': 0.3, 'maxOutputTokens': 4096}
    }).encode('utf-8')
    req = urllib.request.Request(url, data=payload, headers={'Content-Type': 'application/json'})
    for attempt in range(MAX_RETRIES):
        try:
            with urllib.request.urlopen(req, timeout=60) as res:
                data = json.loads(res.read().decode('utf-8'))
                text = data.get('candidates', [{}])[0].get('content', {}).get('parts', [{}])[0].get('text', '')
                if text:
                    return text
                raise Exception('Empty response')
        except urllib.error.HTTPError as e:
            body = e.read().decode('utf-8', errors='replace')
            if e.code == 429 or 'quota' in body.lower():
                wait = (attempt + 1) * 15
                print(f'  Rate limited, waiting {wait}s...')
                time.sleep(wait)
                continue
            raise Exception(f'HTTP {e.code}: {body[:200]}')
        except urllib.error.URLError as e:
            if attempt < MAX_RETRIES - 1:
                time.sleep(5)
                continue
            raise
    raise Exception('Max retries exceeded')

def main():
    questions = load_questions()
    progress = load_progress()
    # Q1 already has detailed_explanation in original data
    to_do = [q for q in questions if str(q['num']) not in progress and (q['num'] == 1 and q.get('detailed_explanation') or q['num'] != 1)]
    # Actually: skip Q1 (already has it), skip ones in progress
    to_do = []
    for q in questions:
        qn = str(q['num'])
        if qn in progress:
            continue
        if q['num'] == 1 and q.get('detailed_explanation'):
            continue
        to_do.append(q)

    print(f'残り{len(to_do)}問を生成 (推定{len(to_do)*WAIT//60}分)')

    for i, q in enumerate(to_do):
        try:
            print(f'[{i+1}/{len(to_do)}] Q{q["num"]}...', end='', flush=True)
            text = call_gemini(build_prompt(q))
            progress[str(q['num'])] = text
            save_progress(progress)
            print(f' OK ({len(text)}字)')
            time.sleep(WAIT)
        except KeyboardInterrupt:
            print('\n中断。再実行で続きから。')
            break
        except Exception as e:
            print(f' ERROR: {e}')
            time.sleep(60)

    # Apply to data.js
    print('data.js に反映中...')
    applied = 0
    for q in questions:
        qn = str(q['num'])
        if qn in progress:
            q['detailed_explanation'] = progress[qn]
            applied += 1
    save_data_js(questions)
    missing = sum(1 for q in questions if not q.get('detailed_explanation'))
    print(f'{applied}問反映完了。残り{missing}問。')

if __name__ == '__main__':
    main()

File: generate-explanations.js
#!/usr/bin/env node
/**
 * Gemini APIを使って全問題の詳細解説を生成し、data.jsに保存するスクリプト
 *
 * 使い方:
 *   node generate-explanations.js YOUR_GEMINI_API_KEY
 *
 * オプション:
 *   --start=N     N番目の問題から開始（デフォルト: 0）
 *   --model=MODEL モデル指定（デフォルト: gemini-2.0-flash-lite）
 *   --delay=MS    リクエスト間の待機時間ms（デフォルト: 5000）
 *
 * 例:
 *   node generate-explanations.js AIzaSy... --start=100 --delay=6000
 *
 * 途中で中断しても、進捗は data_with_details.js に保存されるので
 * --start で再開できます。
 */

const fs = require('fs');
const https = require('https');
const path = require('path');

// ===== 引数解析 =====
const args = process.argv.slice(2);
const API_KEY = args.find(a => !a.startsWith('--'));
if (!API_KEY) {
    console.error('使い方: node generate-explanations.js YOUR_API_KEY [--start=N] [--model=MODEL] [--delay=MS]');
    process.exit(1);
}

const START_INDEX = parseInt((args.find(a => a.startsWith('--start=')) || '').split('=')[1]) || 0;
const MODEL = (args.find(a => a.startsWith('--model=')) || '').split('=')[1] || 'gemini-2.0-flash-lite';
const DELAY_MS = parseInt((args.find(a => a.startsWith('--delay=')) || '').split('=')[1]) || 5000;

// ===== データ読み込み =====
const dataPath = path.join(__dirname, 'data.js');
const outputPath = path.join(__dirname, 'data_with_details.js');
const content = fs.readFileSync(dataPath, 'utf-8');
const start = content.indexOf('[');
const end = content.lastIndexOf(']') + 1;
const questions = JSON.parse(content.substring(start, end));

// 既存の進捗を読み込む
let existingData = null;
if (fs.existsSync(outputPath)) {
    try {
        const existing = fs.readFileSync(outputPath, 'utf-8');
        const s = existing.indexOf('[');
        const e = existing.lastIndexOf(']') + 1;
        existingData = JSON.parse(existing.substring(s, e));
        console.log(`既存の進捗ファイルを検出: ${existingData.filter(q => q.detailed_explanation).length}問 生成済み`);
        // Merge existing detailed_explanations
        existingData.forEach((eq, i) => {
            if (eq.detailed_explanation && !questions[i].detailed_explanation) {
                questions[i].detailed_explanation = eq.detailed_explanation;
            }
        });
    } catch (e) {
        console.log('既存ファイルの読み込みに失敗、最初から開始します');
    }
}

console.log(`\n全${questions.length}問中、問題${START_INDEX + 1}から開始`);
console.log(`モデル: ${MODEL}`);
console.log(`リクエスト間隔: ${DELAY_MS}ms\n`);

// ===== API呼び出し =====
function callGemini(prompt) {
    return new Promise((resolve, reject) => {
        const url = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent?key=${API_KEY}`;
        const body = JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: {
                temperature: 0.3,
                maxOutputTokens: 4096
            }
        });

        const urlObj = new URL(url);
        const options = {
            hostname: urlObj.hostname,
            path: urlObj.pathname + urlObj.search,
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(body)
            }
        };

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (res.statusCode === 429) {
                        reject(new Error('RATE_LIMIT'));
                        return;
                    }
                    if (res.statusCode !== 200) {
                        reject(new Error(`HTTP ${res.statusCode}: ${json?.error?.message || data.substring(0, 200)}`));
                        return;
                    }
                    const text = json?.candidates?.[0]?.content?.parts?.[0]?.text;
                    if (!text) {
                        reject(new Error('空の応答'));
                        return;
                    }
                    resolve(text);
                } catch (e) {
                    reject(e);
                }
            });
        });
        req.on('error', reject);
        req.write(body);
        req.end();
    });
}

function buildPrompt(q) {
    let choicesText = '';
    if (q.choices && q.choices.length > 0) {
        choicesText = q.choices.map(c => `${c.letter}. ${c.text}`).join('\n');
    }

    return `あなたはCompTIA Security+ (SY0-701) の試験対策の専門家です。
以下の問題について、初心者にもわかるように詳しく日本語で解説してください。

【問題】
${q.question}

【選択肢】
${choicesText}

【正解】
${q.answer}

以下の構成で解説してください：

## 1. なぜ「${q.answer}」が正解なのか？
正解の根拠と背景知識を、具体例を交えて丁寧に説明してください。

## 2. 各不正解の選択肢が間違いである理由
各不正解選択肢について、なぜ不正解なのかを具体的に説明してください。

## 3. この問題に関連する重要な概念やキーワード
試験に出る可能性のある関連用語や概念を、それぞれ簡単な説明付きでリストアップしてください。

## 4. 試験で覚えておくべきポイント
この問題から学べる試験対策のポイントを箇条書きでまとめてください。

わかりやすく、体系的に解説してください。`;
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function saveProgress() {
    const output = 'const QUESTIONS = ' + JSON.stringify(questions, null, 2) + ';\n';
    fs.writeFileSync(outputPath, output, 'utf-8');
}

// ===== メイン処理 =====
async function main() {
    let generated = questions.filter(q => q.detailed_explanation).length;
    let errors = 0;

    for (let i = START_INDEX; i < questions.length; i++) {
        const q = questions[i];

        // 既に生成済みならスキップ
        if (q.detailed_explanation) {
            console.log(`[${i + 1}/${questions.length}] Q${q.num} ✓ スキップ（生成済み）`);
            continue;
        }

        const prompt = buildPrompt(q);
        let retries = 0;
        const maxRetries = 5;

        while (retries < maxRetries) {
            try {
                process.stdout.write(`[${i + 1}/${questions.length}] Q${q.num} 生成中...`);
                const result = await callGemini(prompt);
                q.detailed_explanation = result;
                generated++;
                console.log(` ✓ (${generated}問完了)`);

                // 10問ごとに保存
                if (generated % 10 === 0) {
                    saveProgress();
                    console.log(`  → 進捗を保存しました`);
                }
                break;
            } catch (err) {
                if (err.message === 'RATE_LIMIT') {
                    retries++;
                    const wait = DELAY_MS * retries * 2;
                    console.log(` ⏳ レート制限、${wait / 1000}秒待機... (${retries}/${maxRetries})`);
                    await sleep(wait);
                } else {
                    console.log(` ✗ エラー: ${err.message}`);
                    errors++;
                    break;
                }
            }
        }

        if (retries >= maxRetries) {
            console.log(`  ✗ Q${q.num}: リトライ上限に達しました。スキップします。`);
            errors++;
        }

        // 通常のリクエスト間隔
        await sleep(DELAY_MS);
    }

    // 最終保存
    saveProgress();

    console.log(`\n===== 完了 =====`);
    console.log(`生成済み: ${generated}/${questions.length}`);
    console.log(`エラー: ${errors}`);
    console.log(`出力: ${outputPath}`);

    if (generated === questions.length) {
        console.log(`\n全問完了！ 以下のコマンドでdata.jsを置き換えてください:`);
        console.log(`  cp ${outputPath} ${dataPath}`);
    } else {
        console.log(`\n未完了の問題があります。以下で再開できます:`);
        const nextUnfinished = questions.findIndex((q, idx) => idx >= START_INDEX && !q.detailed_explanation);
        if (nextUnfinished >= 0) {
            console.log(`  node generate-explanations.js ${API_KEY.substring(0, 8)}... --start=${nextUnfinished}`);
        }
    }
}

main().catch(err => {
    console.error('致命的なエラー:', err);
    saveProgress();
    console.log('進捗を保存しました。');
    process.exit(1);
});
